src/asl.ts

```diff
@@ -1495,37 +1668,58 @@ export class ASL {
       addState: (state) => {
         states.push(state);
       },
-      // normalizes the output into a json path, creating a Pass state to turn a constant into a jsonPath if needed.
-      normalizeOutputToJsonPath: () => {
-        if (ASLGraph.isJsonPath(output)) {
-          return output;
-        } else {
-          const heap = this.newHeapVariable();
-
-          states.push({
-            ...ASLGraph.passWithInput(
-              {
-                Type: "Pass",
-                ResultPath: heap,
-                Next: ASLGraph.DeferNext,
-              },
-              output
-            ),
-            node: expr,
-          });
-
-          return {
-            jsonPath: heap,
-          };
-        }
-      },
     });
 
     // return the value generated by the handler and any intermediate states generated
     // by eval, `addState`, or `normalizeOutputToJsonPath`.
     return [handlerOutput, states];
   }
 
+  /**
+   * Returns the states required to normalize any {@link ASLGraph.Output} into a {@link ASLGraph.JsonPath}.
+   *
+   * * If the output was a {@link ASLGraph.LiteralValue}, a new state will be added that turns the literal into a json path.
+   * * If the output was a {@link ASLGraph.JsonPath}, the output is returned.
+   * * If the output was a {@link ASLGraph.ConditionOutput}, a new {@link Choice} state will turn the conditional into a boolean
+   *   and return a {@link ASLGraph.JsonPath}.
+   */
+  private normalizeOutputToJsonPath(
+    output: ASLGraph.Output,
+    node?: FunctionlessNode
+  ):
+    | ((ASLGraph.OutputState | ASLGraph.OutputSubState) & {
+        output: ASLGraph.JsonPath;
+      })
+    | ASLGraph.JsonPath {
+    if (ASLGraph.isJsonPath(output)) {
+      return output;
+    } else {
+      return this.assignValue(node, output);
+    }
+  }
+
+  /**
+   * returns the states required to normalize a {@link ASLGraph.Output} into a {@link ASLGraph.JsonPath} or {@link ASLGraph.LiteralValue}.
+   *
+   * * If the output is a {@link ASLGraph.LiteralValue}, the output is returned as is.
+   * * If the output was a {@link ASLGraph.JsonPath}, the output is returned as is.
+   * * If the output was a {@link ASLGraph.ConditionOutput}, a new {@link Choice} state will turn the conditional into a boolean
+   *   and return a {@link ASLGraph.JsonPath}.
+   */
+  private normalizeConditionToJsonPath(
+    output: ASLGraph.Output,
+    node?: FunctionlessNode
+  ):
+    | (ASLGraph.OutputSubState & { output: ASLGraph.JsonPath })
+    | ASLGraph.JsonPath
+    | ASLGraph.LiteralValue {
+    if (ASLGraph.isJsonPath(output) || ASLGraph.isLiteralValue(output)) {
+      return output;
+    } else {
+      return this.conditionState(node, output.condition);
+    }
+  }
+
```

Comment:
This seems to handle more than `ConditionToJsonPath`.

Reply:
Updated `normalizeOutputToJsonPathOrLiteralValue`.